enhancements/ingress/bind-options.md

```diff
@@ -0,0 +1,268 @@
+---
+title: bind-options
+authors:
+  - "@m-yosefpor"
+reviewers:
+  -
+approvers:
+  -
+creation-date: 2021-08-04
+last-updated: 2021-08-04
+status: implementable
+see-also:
+  - https://github.com/openshift/cluster-ingress-operator/issues/633
+  - https://github.com/openshift/api/issues/964
+replaces:
+superseded-by:
+---
+
+# Ingress Bind Options
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement extends the IngressController API to allow the user to
+specify bindOptions in HostNetwork strategy and enables them to run multiple
+instances of ingresses on the same node in the HostNetwork strategy, with
+mitigating port binding conflicts.
+
+## Motivation
+
+When using HostNetwork strategy for ingressControllers, the default 80, 443, 10443, 10444, 1936 ports on the host are binded on HAProxy for http, https, no_sni, sni and stats correspondingly. However, those ports might be occupied by other processes (such as another set of ingressControllers), which makes it impossible to run multiple set of ingressControllers on the same nodes with HostNetwork strategy. In OKD3.11, it was possible to listen on custom host ports via setting these env vars in Routers DeploymentConfig, however there is not any options to specify custom ports in `IngressController.operator.openshift.io/v1` object right now.
+
+Having multiple set of ingressControllers is useful for router sharding, routers with different policies (public, private), routers with different configuration options, etc. Running in HostNetwork is a strict requirement in some scenarios (e.g. environemtns with custom PBR rules). Also there might be some performance considerations to run routers with HostNetwork strategy.
```

Comment:
Policy Based Routing rules, e.g. when we have source-ip based routing rules such as
```
ip rule add from 192.168.1.0/24 table mytable prio 1
```

In our case, we have multiple default GWs, and each packet should move back from the interface which has arrived to the node. In such cases, the pod routing rules and tables should match the ones on the host, otherwise the traffic is not routed correctly. Although If there are only static route records in main routing table, network namespaces works fine, but with PBR rules, the pods need to run on the hostNetwork, unless someone wants to do hacky workaround like adding the rules with initContainers (which is also not possible in-case of ingress pods as the deployment is created by CIO not cluster-admins)

For performance considerations, I meant some users want to run the ingress pods in hostNetwork since it has a better performance (probably there should be better wording for that instead of `performance considerations`)

better perfomance of hostNetwork routers has been shown in the OKD docs benchmarks (https://docs.okd.io/latest/scalability_and_performance/routing-optimization.html).


Shall I mention the above details in proposal too?

Reply:
No, I think it's fine like this.

About the performance it wasn't clear to me whether you meant performance benefits or penalties. Yes, may be a different wording or even the link to OKD doc would make the phrase absolutely clear.

Reply:
fixed.